src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
+you'd want to do a null check to see whether `this` is null:
+
+**Example:**
+
+{:.bad}
+```dart
+extension E on int? {
+  int? get i {
+    return this != null;
+  }
+}
+```
+
+**Message:**
+
+```nocode
+`this` can't be promoted. 
+```
+
+**Solution:**
+
+Create a local variable to hold the value of `this`, then perform the null check.
+
+{:.good}
+```dart
+extension E on int? {
+  int? get i {
+    final self = this;
+    return self != null;
+  }
+}
+```
+
+### Only private fields can be promoted {#private}
+
+**The cause:**
+You're trying to promote a field, but the field is not private.
+
+It’s possible for other libraries in your program
+to override public fields with a getter. Because
+[getters might not return a stable value](#not-field),
+and the compiler can't know what other libraries are doing,
+non-private fields cannot be promoted.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  final int? n; // CONTEXT
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // ERROR
+  }
+}
+```
+
+Here, the compiler produces an error message
+explaining that the `+` operator can't be called on `n`
+because it is potentially null.
+
+**Message:**
+
+```nocode
+'n' refers to a public field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Making the field private lets the compiler be sure that no outside libraries
+could possibly override its value, so it's safe to promote.
+
+{:.good}
+```dart
+class C {
+  final int? _n;
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // OK
+  }
+}
+```
+
+### Only final fields can be promoted {#final}
+
+**The cause:**
+You're trying to promote a field, but the field is not final.
+
+To the compiler, non-final fields could, in principle,
+be modified any time between the time
+they’re tested and the time they’re used.
+So it's not safe for the compiler to promote a non-final nullable type
+to a non-nullable type.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  int? _mutablePrivateField; // CONTEXT
+  Example(this._mutablePrivateField);
+
+  f() {
+    if (_mutablePrivateField != null) {
+      int i = _mutablePrivateField; // ERROR
+    }
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'mutablePrivateField' refers to a non-final field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Make the field `final`
+
+{:.good}
+```dart
+class Example {
+  int? final _immutablePrivateField; 
+  Example(this._immutablePrivateField);
+
+  f() {
+    if (_mutablePrivateField != null) {
+      int i = _immutablePrivateField; // OK
+    }
+  }
+}
+```
+
+### Getters can't be promoted {#not-field}
+
+**The cause:** You're trying to promote a getter,
+but only instance fields can be promoted, not getters. 
+
+The compiler has no way to guarantee that a getter returns the same result every time.
+Because their stability can't be confirmed, getters are not safe to promote.
+
+**Example:**
+
+{:.bad}
+```dart
+abstract class C {
```

Comment:
We're planning to make a breaking change to the language in 3.3 that's going to render this example invalid. (Details, if you're curious, are here: https://github.com/dart-lang/language/issues/3328#issuecomment-1792511446). We should probably tweak this example so that the breaking change won't affect it.

Here's an example you can use:
```dart
import 'dart:math';
abstract class C {
  int? get _i => Random().nextBool() ? 123 : null; // CONTEXT
}

f(C c) {
  if (c._i != null) {
    pint(c._i.isEven); // ERROR
  }
}
```

And the corresponding "good" code would be:
```dart
import 'dart:math';
abstract class C {
  int? get _i => Random().nextBool() ? 123 : null;
}

f(C c) {
  final i = c._i;
  if (i != null) {
    print(i.isEven);
  }
}
```

(Note: I switched to printing whether the integer is even rather than trying to assign it to an `int` variable, just so that I wouldn't have to name the new variable `j`).

Reply:
Lmk if the note I added on [line 361](https://github.com/dart-lang/site-www/pull/5246/files#diff-68e8ac8e2954d6b7811e6c6a95e2fe171417e5fbd368a6f03f30f6ac9645fc1fR361-R367) sounds good!

Reply:
Yeah, that looks totally reasonable.